cmds/shutdown/shutdown_test.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"os"
+	"os/exec"
+	"syscall"
+	"testing"
+
+	"golang.org/x/sys/unix"
+)
+
+var tests = []struct {
+	args    []string
+	retCode int
+}{
+	// Too many args is an error.
+	{[]string{"halt", "police"}, 1},
+	{[]string{"halt"}, 2},
+	{[]string{"-h"}, 2},
+	// No args means halt.
+	{[]string{}, 2},
+	{[]string{"reboot"}, 3},
+	{[]string{"-r"}, 3},
+	{[]string{"suspend"}, 4},
+	{[]string{"-s"}, 4},
+}
+
+func TestShutdown(t *testing.T) {
+	for i, tt := range tests {
+		var retCode int
+		c := exec.Command(os.Args[0], append([]string{"-test.run=TestHelperProcess", "--"}, tt.args...)...)
+		c.Env = []string{"GO_WANT_HELPER_PROCESS=1"}
+		t.Logf("Run %v", c)
+		_, err := c.Output()
+		t.Logf("err %v", err)
+		if err != nil {
+			exitErr, ok := err.(*exec.ExitError)
+			if !ok {
+				t.Errorf("%d. Error running shutdown: %v", i, err)
+				continue
+			}
+			retCode = exitErr.Sys().(syscall.WaitStatus).ExitStatus()
+		}
+		if retCode != tt.retCode {
+			t.Errorf("%d. Want: %d; Got: %d", i, tt.retCode, retCode)
+		}
+	}
+}
+
+func TestHelperProcess(t *testing.T) {
+	if os.Getenv("GO_WANT_HELPER_PROCESS") != "1" {
```

Comment:
Why make this conditional if you always use it in the tests?

Reply:
it's a technique used in go packages.

we run the test explicitly via test.run. So it needs to be visible. 

But then go test would run it. So we add this env to indicate that it's running as a helper process. If it's not running as a helper process, we skip the test.